File: src/main/java/view/TileBuffer.java
package view;

import org.jline.terminal.Terminal;

class TileBuffer {
    private Terminal mTerminal;
    private Tile[][] mBuffer;

    // lifetime
    TileBuffer(Terminal terminal) {
        mTerminal = terminal;
    }

    // commands
    void initialize(int length) {
        mBuffer = new Tile[length][length];
    }

    void set(int x, int y, Tile tile) {
        mBuffer[y][x] = tile;
    }

    void draw() {
        if (!isInitialized()) {
            throw new Errors.NoBoard();
        }

        for (Tile[] row : mBuffer) {
            int i = 0;

            for (Tile tile : row) {
                mTerminal.writer().print(tile.render());
                mTerminal.writer().print(i++ == row.length - 1 ? '\n' : ' ');
            }
        }

        mBuffer = null;
    }

    // queries
    boolean isInitialized() {
        return mBuffer != null;
    }

    boolean contains(int x, int y) {
        return y < mBuffer.length && x < mBuffer[y].length;
    }
}

File: Sources/Onitama/view/TileBuffer.swift
final class TileBuffer {
  private let mTerminal: Terminal
  private var mBuffer: [[Tile]]?

  // -- lifetime --
  init(terminal: Terminal) {
    mTerminal = terminal
  }

  // -- commmands --
  func initialize(size: Int) {
    mBuffer = Array.grid(repeating: Tile(), count: size)
  }

  func set(_ point: Point, tile: Tile) {
    guard let buffer = mBuffer else {
      fatalError("buffer was not initialized")
    }

    if point.x >= buffer.count || point.y >= buffer.count {
      fatalError("buffer does not contain \(point)")
    }

    mBuffer?[point.y][point.x] = tile
  }

  func draw() {
    guard let buffer = mBuffer else {
      fatalError("buffer was not initialized")
    }

    for row in buffer {
      for (i, tile) in row.enumerated() {
        mTerminal.write(tile.render())
        mTerminal.write(i == row.count - 1 ? "\n" : " ")
      }
    }

    mBuffer = nil
  }
}
